### PR TITLE
Require -w/--threads or MZ_THREADS

### PR DESCRIPTION
**Pull request by cuongdo**
To educate users and make -w / --threads discoverable, we require that
the user specify the number of worker threads when running a release
binary.

This isn't a requirement for debug builds (--debug / MZ_DEV), because
performance is unlikely to be a primary consideration in that situation.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/2447)

### CODE REVIEW

File: src/materialized/bin/materialized.rs [hunk 1]
```diff
@@ -158,13 +158,24 @@ fn run() -> Result<(), failure::Error> {
...
+    if cfg!(debug_assertions) && !debug_opt_in {
         bail!(
             "refusing to run dev (unoptimized) binary without explicit opt-in\n\
              hint: Pass the '--dev' option or set the MZ_DEV environment variable to opt in.\n\
              hint: Or perhaps you meant to use a release binary?"
         );
     }
+    if !popts.opt_present("threads") && !debug_opt_in {
+        bail!(
+            "missing required number of worker threads '--threads'\n\
```
Comment (ruchirK) [inline src/materialized/bin/materialized.rs]: we might want to say something more explicit about it being required for release builds?

e.g. "missing required number of worker threads for release builds '--threads'"

File: src/materialized/bin/materialized.rs [hunk 2]
```diff
@@ -158,13 +158,24 @@ fn run() -> Result<(), failure::Error> {
...
+             hint: As a starting point, set the number of threads to half of the number of\n\
+             cores on your system. Then, further adjust based on your performance needs.\n\
+             hint: Debug builds default to one worker thread."
```
Comment (ruchirK) [inline src/materialized/bin/materialized.rs]: dev builds to follow the convention from above

**Comment by ruchirK** [comment]
I think line 82 in this file has the potential to be confusing for people

```
 82         "number of per-process worker threads (default 1)",
```

the --help output will disagree with what happens when you run a release build. not totally sure of the best way to fix but you could use compiler directives for debug_assertions to specify what the hint string is?

**Comment by benesch** [comment]
I'd much prefer that this didn't differ between debug and release builds. The behavior of a debug build should be identical to the behavior of a release build. (IMO it's problematic that we've introduced the `--dev` flag, but I lost that argument.)

So I propose the following behavior. Either the `--threads` or the `MZ_THREADS` environment variable must be specified, for both dev and release builds. If the command line argument and environment variable are both specified, then the command line argument takes precedence over the environment variable.

**Comment by cuongdo** [comment]
Thanks for the feedback. I'll revise and let you know when it's ready for another review.

File: src/materialized/bin/materialized.rs [hunk 3]
```diff
@@ -151,7 +146,24 @@ fn run() -> Result<(), failure::Error> {
 
     let log_file = popts.opt_str("log-file");
     let max_increment_ts_size = popts.opt_get_default("batch-size", 10000_i64)?;
-    let threads = popts.opt_get_default("threads", 1)?;
+
+    let threads = match popts.opt_get::<usize>("threads")? {
+        Some(val) => val,
+        None => match env::var_os("MZ_THREADS") {
+            Some(val) => val.into_string().unwrap().parse()?,
```
Comment (benesch) [inline src/materialized/bin/materialized.rs]: Unwrapping here will panic if `MZ_THREADS` contains non-Unicode characters. Instead you'll want to use `env::var("MZ_THREADS")`, and match on the error, ignoring `VarError::NotPresent` and propagating `VarError::NotUnicode`.

Comment (cuongdo) [reply src/materialized/bin/materialized.rs]: Great catch!

**Pull request merged**